Split Batcher logic with OpenGl implementation

karupayun opened the pull request:
Following decisions in go/stadia-orbit-testing-ui-elements, we want to
refactor Batcher functionality to make it easier to mock. In this PR we
are creating OpenGlBatcher (first step in http://b/225173189).

In addition we cleaned includes and refactor virtual methods to not have
default parameters.

Next step: Create PrimitiveAssembler.

beckerhe commented on src/OrbitGl/BatcherImpl.h:
```diff
@@ -0,0 +1,102 @@
+ public:
+  BatcherImpl(BatcherId batcher_id, PickingManager* picking_manager = nullptr)
+      : Batcher(batcher_id, picking_manager) {}
+  [[nodiscard]] std::vector<float> GetLayers() const override;
+  void DrawLayer(float layer, bool picking) const override;
```
After seeing the code, I could also imagine having multi level inheritance. Something like that:

- `Batcher` (inherits from nothing) declares a bunch of pure virtual functions for adding primitives
- `DrawingBatcher` (inherits from `Batcher`) declares `GetLayers` and `DrawLayer` as pure virtual functions
- `DrawingBatcherImpl` (inherits from `DrawingBatcher`) implements all of it. This class could be declared in a cpp-file and be a proper implementation detail.

For your tests you can mock or fake `Batcher`. The `CaptureWindow` would get a pointer to `DrawingBatcher` because it needs `DrawLayer`, but all the rest probably only needs a pointer to `Batcher`.

reichlfl replied:
Additional thoughts: You mention you want to rename this to `BatcherInterface`, but if I understand correctly, this will still include implementation of non-private methods? In that case it's not an interface. Did you consider composition over inheritance? For example:

* `BatcherInterface` declares only the virtual functions needed to add primitives, but as public API
* `OpenGlBatcher` implements this interface
* Create a new class `PrimitiveAssembler` (... better names welcome) that receives a pointer to a `BatcherInterface` and is responsible of assembling complex elements (circles, shaded boxes, ...) using the `BatcherInterface` methods

karupayun replied:
Really thanks for the suggestions. A few thoughts. I feel:

- The goal of this PR is isolate OpenGl related methods with the rest in Batcher (Thanks Florian for the better name, I should use OpenGlBatcher!).
- Henning is trying to additionally isolate drawing primitives with adding them.
- Florian is trying to additionally isolate proper primitives with additional methods to add them. 

I found both proposals correct, I was a bit lazy to think about all this refactors in this PR, but I should have a clearly defined long-time proposal before merging this, thanks for making me realize that.

If we are worried about the long-term picture, I will include merging TextRenderer and Batcher. For example, in Florian's proposal, PrimitiveAssembler could have both pointer of BatcherInterface and TextRendererInterface.

I will update the doc before going forward with this PR.

reichlfl replied:
Fair enough to not do this all in a single PR, this can absolutely be approached iteratively and I think this is the first step in the right direction.

beckerhe reviewed (commented):
Since you're already changing the names, would you consider moving Batcher and BatcherInterface into the proper namespace?

reichlfl commented on src/OrbitGl/Batcher.h:
```diff
@@ -97,9 +44,8 @@ into layers formed by equal z-coordinates. Each layer can then be drawn seperate
-NOTE: The Batcher assumes x/y coordinates are in pixels and will automatically round those
-down to the next integer in all Batcher::AddXXX methods. This fixes the issue of primitives
-"jumping" around when their coordinates are changed slightly.
+NOTE: Batcher has a few pure virtual functions that has to be implemented: A few AddInternalMethods,
```
... that **have** to be implemented ....

antonrohr reviewed (commented):
Thanks for including me on this PR. I do not have comments right now, but maybe after you addressed the other comments. 

Regarding:
> Since you're already changing the names, would you consider moving Batcher and BatcherInterface into the proper namespace?

I think this is a good idea and your upcoming renaming PR is a good place for that

beckerhe commented:
> Thanks for including me on this PR. I do not have comments right now, but maybe after you addressed the other comments.
> 
> Regarding:
> 
> > Since you're already changing the names, would you consider moving Batcher and BatcherInterface into the proper namespace?
> 
> I think this is a good idea and your upcoming renaming PR is a good place for that

Oh, I think I misunderstood that. Are you planning to do the renaming in a separate PR?

karupayun commented on src/OrbitGl/BatcherImpl.h:
```diff
@@ -0,0 +1,102 @@
+  orbit_containers::BlockChain<Color, 2 * NUM_LINES_PER_BLOCK> picking_colors_;
+};
+
+struct BoxBuffer {
+  void Reset() {
+    boxes_.Reset();
+    colors_.Reset();
+    picking_colors_.Reset();
+  }
+
+  static const int NUM_BOXES_PER_BLOCK = 64 * 1024;
+  orbit_containers::BlockChain<Box, NUM_BOXES_PER_BLOCK> boxes_;
+  orbit_containers::BlockChain<Color, 4 * NUM_BOXES_PER_BLOCK> colors_;
+  orbit_containers::BlockChain<Color, 4 * NUM_BOXES_PER_BLOCK> picking_colors_;
+};
+
+struct TriangleBuffer {
+  void Reset() {
+    triangles_.Reset();
+    colors_.Reset();
+    picking_colors_.Reset();
+  }
+
+  static const int NUM_TRIANGLES_PER_BLOCK = 64 * 1024;
+  orbit_containers::BlockChain<Triangle, NUM_TRIANGLES_PER_BLOCK> triangles_;
+  orbit_containers::BlockChain<Color, 3 * NUM_TRIANGLES_PER_BLOCK> colors_;
+  orbit_containers::BlockChain<Color, 3 * NUM_TRIANGLES_PER_BLOCK> picking_colors_;
+};
+
+struct PrimitiveBuffers {
+  void Reset() {
+    line_buffer.Reset();
+    box_buffer.Reset();
+    triangle_buffer.Reset();
+  }
+
+  LineBuffer line_buffer;
+  BoxBuffer box_buffer;
+  TriangleBuffer triangle_buffer;
+};
```
> Since you're already changing the names, would you consider moving Batcher and BatcherInterface into the proper namespace?

Will add it for OpenGlBatcher. After changing Batcher to BatcherInterface I will also include it in orbit_gl namespace.

karupayun commented:
> Since you're already changing the names, would you consider moving Batcher and BatcherInterface into the proper namespace?

Will add it for OpenGlBatcher. After changing Batcher to BatcherInterface I will also include it in orbit_gl namespace.

karupayun commented:
Create a bug with steps: http://b/225173189.

Renamed to OpenGlBatcher, added a few namespace, change a comment. 

Asking for a review now that we have a clearer path.

antonrohr reviewed (approved):
LGTM, a second review other than mine might be a good idea, I might have missed something

beckerhe commented on src/OrbitGl/OpenGlBatcher.h:
```diff
@@ -0,0 +1,109 @@
+};
+
+}  // namespace
+/**
+Implements internal methods to collects primitives to be rendered at a later point in time.
+
+NOTE: The OpenGlBatcher assumes x/y coordinates are in pixels and will automatically round those
+down to the next integer in all Batcher::AddXXX methods. This fixes the issue of primitives
+"jumping" around when their coordinates are changed slightly.
+**/
```
Nit: We tend to use `//` instead of `/* */`

beckerhe replied:
Also an empty line above the comment would be great

karupayun replied:
Yes, it was copy pasted directly from Batcher.

beckerhe commented on src/OrbitGl/OpenGlBatcher.h:
```diff
@@ -0,0 +1,109 @@
+**/
+class OpenGlBatcher : public Batcher {
+ public:
+  OpenGlBatcher(BatcherId batcher_id, PickingManager* picking_manager = nullptr)
```
`explicit`

beckerhe commented on src/OrbitGl/OpenGlBatcher.h:
```diff
@@ -0,0 +1,109 @@
+  void DrawLayer(float layer, bool picking) const override;
+
+ protected:
+  std::unordered_map<float, PrimitiveBuffers> primitive_buffers_by_layer_;
```
Why does this need to be `protected`? Couldn't this be `private`?

karupayun replied:
It's used internally in batcher test. Do you prefer me to provide a protected method to get it?

beckerhe commented on src/OrbitGl/OpenGlBatcher.h:
```diff
@@ -0,0 +1,109 @@
+
+namespace orbit_gl {
+
+namespace {
```
You shouldn't have an anonymous namespace in a header. If you want to have these types internal put it into the namespace `orbit_gl_internal`.

beckerhe reviewed (approved):
Otherwise LGTM

reichlfl reviewed (approved):
Please address Henning's comments, otherwise LGTM!
